### PR TITLE
Account balance model

### PR DESCRIPTION
**Pull request by make-it-git**
## Overview

| Q                 | A
| ----------------- | ---
| Bugfix?           | no
| New feature?      | yes
| Breaking Changes? | no
| Fixed issues      | #54 (partially)

## The problem

SageOne is very inconsistent in methods/response so I added dummy model for response from getAccountBudgets()

Also all() method requires POST (not GET).

Notify the following people: @darrynten

### CODE REVIEW

File: src/BaseModel.php [hunk 1]
```diff
@@ -53,6 +53,19 @@
     ];
 
     /**
+     * Features HTTP methods
+     * Not all models follow same conventions like GET for all()
+     * Example AccountBalance all() requires POST method
```
Comment (darrynten) [inline src/BaseModel.php:58]: I think this is very important, can you please add a note to the README in the relevant section?

Strange inconsistencies :)

File: src/BaseModel.php [hunk 2]
```diff
@@ -150,7 +163,7 @@ public function all()
             $this->throwException(ModelException::NO_GET_ALL_SUPPORT);
         }
 
-        $results = $this->request->request('GET', $this->endpoint, 'Get');
+        $results = $this->request->request($this->featureMethods['all'], $this->endpoint, 'Get');
```
Comment (darrynten) [inline src/BaseModel.php]: I like the way you solved it, but it's going to be crucial to document it sufficiently in the Example model and in the Readme

**Review by darrynten — CHANGES_REQUESTED** [review]
I really like the way you fixed the problem here.

The only thing we need is for new changes to be documented in Readme and in ExampleModel.

@fergusdixon @kiproping @igorsergiichuk take note of this clever change :)

File: src/BaseModel.php (hunk 2 above)
Comment (make-it-git) [reply src/BaseModel.php]: @darrynten done

Comment (darrynten) [reply src/BaseModel.php]: @make-it-git there are some conflicts in this branch that need to be resolved before we can merge it in

File: src/Models/AccountBalance.php [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        'get' => '',
+        'save' => '',
+        'delete' => ''
+    ];
+
+    /**
+     * $param integer $id
+     * @return ModelCollection A collection of Account Balance
+     */
+    public function getAccountBudgetsById($id)
```
Comment (darrynten) [inline src/Models/AccountBalance.php:124]: Codacy Issue found: [Avoid variables with short names like $id. Configured minimum length is 3.](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=785392)

File: tests/SageOne/Models/AccountBalanceModelTest.php [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        $this->verifyFeatures(AccountBalance::class, [
+            'all' => true, 'get' => false, 'delete' => false, 'save' => false
+        ]);
+    }
+
+    public function testGetAll()
+    {
+        $this->verifyGetAll(AccountBalance::class, function ($results) {
+            $this->assertEquals(2, count($results));
+
```
Comment (darrynten) [inline tests/SageOne/Models/AccountBalanceModelTest.php]: Codacy Issue found: [Whitespace found at end of line](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=785392)

File: tests/SageOne/Models/BaseModelTest.php [hunk 5]
```diff
@@ -558,18 +558,25 @@ protected function verifyInject(string $class, callable $whatToCheck)
      * Verifies that we can load list of models
      *
      * @param string $class Full path to the class
+     * @param string $method HTTP method of call
+     * @param string $requestMock If specified sends data from that file in request
      * @param callable $whatToCheck Verifies fields on result
      */
-    protected function verifyGetAll(string $class, callable $whatToCheck)
+    protected function verifyGetAll(string $class, callable $whatToCheck, string $method = 'GET', string $requestMock = null)
```
Comment (darrynten) [inline tests/SageOne/Models/BaseModelTest.php:565]: Codacy Issue found: [Line exceeds 120 characters; contains 125 characters](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=785392)

File: tests/SageOne/Models/BaseModelTest.php [hunk 6]
```diff
@@ -558,18 +558,25 @@ protected function verifyInject(string $class, callable $whatToCheck)
...
     {
         $className = $this->getClassName($class);
         $path = sprintf('%s/Get', $className);
-        $mockFile = sprintf('%s/GET_%s_Get.json', $className, $className);
+        if ($requestMock) {
+            $responseMock = sprintf('%s/%s_%s_Get_RESP.json', $className, $method, $className);
+        } else {
```
Comment (darrynten) [inline tests/SageOne/Models/BaseModelTest.php]: Codacy Issue found: [The method verifyGetAll uses an else expression. Else is never necessary and you can simplify the code to work without else.](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=785392)

**Comment by darrynten** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- tests/SageOne/Models/BaseModelTest.php  1
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=785392&bid=4951591)

File: src/Models/AccountBalance.php (hunk 3 above)
Comment (darrynten) [reply src/Models/AccountBalance.php:124]: this one is invalid

File: tests/SageOne/Models/BaseModelTest.php (hunk 5 above)
Comment (darrynten) [reply tests/SageOne/Models/BaseModelTest.php:565]: invalid

**Comment by make-it-git** [comment]
@darrynten merge please

**Comment by darrynten** [comment]
@make-it-git there are still change requests on this PR 😄

File: tests/SageOne/Models/BaseModelTest.php (hunk 6 above)
Comment (darrynten) [reply tests/SageOne/Models/BaseModelTest.php]: @make-it-git here

File: tests/SageOne/Models/AccountBalanceModelTest.php (hunk 4 above)
Comment (darrynten) [reply tests/SageOne/Models/AccountBalanceModelTest.php]: @make-it-git here

**Review by darrynten — CHANGES_REQUESTED** [review]
Highlighting the requested changes

**Review by darrynten — APPROVED** [review]
Really great work here!

**Pull request merged**